### ISSUE TITLE
Band Artefacts

### ISSUE THREAD

**Issue by utagawal** [acalcutt/jaxa_AW3D30_to_MBTiles#5]
Downloaded your mbtile terrain 0-12. Thanks for making this available.

Some artefacts are displayed
![image](https://github.com/acalcutt/jaxa_AW3D30_to_MBTiles/assets/16464382/bc916746-ba37-43bf-a340-1c697f4b714f)

Any idea to avoid this ?

**Comment by acalcutt** [comment]
Based on the resampling comparisons at https://github.com/alukach/vector-ocean?tab=readme-ov-file#1-sourcing--preparing-data , I wonder if switching to 'cubic' or 'bilinear' instead of 'cubicspline ' at https://github.com/acalcutt/jaxa_AW3D30_to_MBTiles/blob/main/create_terrainrgb.sh#L23 would make a difference.

**Comment by utagawal** [comment]
Nice catch. Any chance you could try and display samples of the 2 ? With the examples, it seems cubic looks a little bit better

**Comment by neodescis** [comment]
I'm looking into these banding issues myself presently, as I've downloaded the pmtiles archive torrent you've provided, @acalcutt. It seems that providing the resampling argument to `gdalbuildvrt` in addition to `gdalwarp` makes things considerably better, especially at higher latitudes. This is because `gdalbuildvrt` also results in pixel shifting before `gdalwarp` is ever even run, and it defaults to "nearest" if no argument is provided. Here's an example taken from Greenland. The first image is using the pmtiles archive I downloaded, and the second is generated using `gdalbuildvrt -r cubic ...` and `gdalwarp -r cubic ...`.

![Image](https://github.com/user-attachments/assets/1c1f1f2e-3ed5-404a-b2e3-4fe81fce0e37)

![Image](https://github.com/user-attachments/assets/7dfbdaa5-af96-42ab-ba07-64d97967043d)

**Comment by neodescis** [comment]
Here's another generated using `-r cubicspline` for both `gdalbuildvrt` and `gdalwarp`. It's smoother, but seems to lose detail.

![Image](https://github.com/user-attachments/assets/001e8a4a-844c-4b73-9403-a238699cb86a)

**Comment by utagawal** [comment]
Good catch, the rendering is great. did you work on a PR ?

**Comment by neodescis** [comment]
I would be happy to create a simple PR for create_terrainrgb.sh, but I suspect all of the scripts need to be updated. Different scripts also currently have different resample methods, so those might need to be standardized. I think cubic looks the best, and I saw you and @acalcutt coming to that conclusion in another issue's discussion. I also suspect the banding @acalcutt has seen when combining terrain sources has the same root cause, but investigating that is outside my scope for the time being. My current goal is just to get a good JAXA AW3D30 tile set.

**Comment by acalcutt** [comment]
Yes, I did decided on cubic. I did just run this version of the script locally https://github.com/acalcutt/terrain_merged/blob/reorg/JAXA_AW3D30/create_terrainrgb.sh

The only difference there is I usually set no-data to 0, but in this I used -10000 because I wanted to see if it made a difference for merging. I can upload a torrent of that version if you would like to try it.

**Comment by neodescis** [comment]
That looks interesting. I have a few questions:

1. What would a dstnodata of -10000 actually look like if I used it directly? If the ocean drops off into the abyss, I don't think that will work for me, haha.
2. Have you further modified rgbify beyond what's in your fork's master branch? I ask because I don't see the "resampling" argument there. I hadn't even realized rgbify was doing its own resampling after gdalwarp! _edit: never mind, I just found your "merge" branch_
3. You said you're running this locally. What kind of computing power do you have? Generating these tiles is taking me days!

At any rate, I'd be extremely grateful for a download. I'm just not sure about the dstnodata value you mentioned. Also, I probably need the archive in WebP format to save on space, but I think maybe you're already doing that?

**Comment by acalcutt** [comment]
1.) I think it would actually look like a drop, unless you added a water layer over it. i still have to test that. 

2.) I am using the rgbify from the merge branch ( https://github.com/acalcutt/rio-rgbify/tree/merge ) like noted at the top of the script. I did further modify rio rgbify to add that option. I also modified it so it uses much less memory in that branch because it no longer holds all the tiles in memory. I can actually run a few copies at once with that memory change

3.) I have two systems I am using right now. One is a virtual server with 12 cores, 78GB memory, disk just on a sata raid 6, and 1000GB swap on a nvme disk.  the other is just a desktop ryzen system, 6core, raid0 with two 4TB nvme disk, 128GB memory. They can take anywhere for a day to a week plus (France seems to be taking me longer for some reason) depending on the zoom level needed.

I usually try to include swap on a fast disk to make up for lack of memory, but with the rgbify changes this might not be needed as much anymore. I do think having nvme disks helps if possible, since they just much faster read/write than normal hdd's or ssds.

I also notice the lower zoom tiles, 0-1 can take forever to generate. I often see it finish most tiles and one thread left is just sitting trying to make those lower tiles.

**Comment by neodescis** [comment]
In that case, I'll pass on the download... I don't think I'll be able to use it with that no-data value. Thank you though!

It sounds like our generation times aren't far off of each other. However, I've actually been parallelizing across machines as well: doing some zoom levels on one, and some on another. Merging the resulting .mbtiles files is relatively straight-forward and quick, since they are just SQLite tables that need copying over. I also was originally doing things by zoom level to drop bits at the end for compressibility, but like you, I did not like the results. Too much terracing.

I've made the same observation about zoom levels 0-1 taking forever. Zoom level 12 also is not terribly quick because of the sheer number of tiles, but at least I can monitor its progress with SQLite.

**Comment by utagawal** [comment]
Regarding computation time : I opened a new PR to see if the script can be adapted, referencing the work mad by @NicoRio42.
I'd be happy to provide computing power to speed up the generation during the tests.

**Comment by NicoRio42** [comment]
> Yes, I did decided on cubic. I did just run this version of the script locally https://github.com/acalcutt/terrain_merged/blob/reorg/JAXA_AW3D30/create_terrainrgb.sh
> 
> The only difference there is I usually set no-data to 0, but in this I used -10000 because I wanted to see if it made a difference for merging. I can upload a torrent of that version if you would like to try it.

@acalcutt I would be interested to see this version if you can upload it for public download. Thanks for all this work!
